flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/CheckpointFailureManager.java

```diff
@@ -31,18 +31,24 @@
  */
 public class CheckpointFailureManager {
 
-	private final static int UNLIMITED_TOLERABLE_FAILURE_NUMBER = Integer.MAX_VALUE;
+	public final static int UNLIMITED_TOLERABLE_FAILURE_NUMBER = Integer.MAX_VALUE;
+	public final static int UNDEFINED_TOLERABLE_CHECKPOINT_NUMBER = -1;
 
 	private final int tolerableCpFailureNumber;
 	private final FailJobCallback failureCallback;
 	private final AtomicInteger continuousFailureCounter;
 	private final Set<Long> countedCheckpointIds;
 
 	public CheckpointFailureManager(int tolerableCpFailureNumber, FailJobCallback failureCallback) {
-		checkArgument(tolerableCpFailureNumber >= 0,
+		checkArgument(tolerableCpFailureNumber >= 0 || tolerableCpFailureNumber == UNDEFINED_TOLERABLE_CHECKPOINT_NUMBER,
 			"The tolerable checkpoint failure number is illegal, " +
 				"it must be greater than or equal to 0 .");
-		this.tolerableCpFailureNumber = tolerableCpFailureNumber;
+		// set tolerableCpFailureNumber as 0 when passing UNDEFINED_TOLERABLE_CHECKPOINT_NUMBER.
+		if (tolerableCpFailureNumber == UNDEFINED_TOLERABLE_CHECKPOINT_NUMBER) {
```

Comment:
I wonder if we should introduce and handle the `UNDEFINED_TOLERABLE_CHECKPOINT_NUMBER` in this class. It seems to me that this is more an implementation detail of how the configuration is determined and propagated. If possible, I would suggest to resolve this already before at the time when the configuration is determined and not in the constuctor of this class. I would suggest to pass only "valid" and resolved configuration here.

Reply:
I agree to make `tolerableCheckpointFailureNumber` valid before `CheckpointFailureManager` or `CheckpointCoordinatorConfiguration`. But I still want to introduce `UNDEFINED_TOLERABLE_CHECKPOINT_NUMBER` at least in `CheckpointConfig` to judge whether `setFailOnCheckpointingErrors` would take effect. 
The only left problem is when we want to call `CheckpointConfig#getTolerableCheckpointFailureNumber`. If that field was already `-1`, shall we set the `tolerableCheckpointFailureNumber` as `0` or just return `0` without setting that. I prefer to return `0` but leave `tolerableCheckpointFailureNumber` still as `-1` to avoid such side effects. What do you think?

Reply:
Yes, your suggestion sounds good and I am also not against this in `CheckpointConfig`, I just think `CheckpointFailureManager` constructor is a bit late.